cachito/web/models.py

```diff
@@ -208,10 +208,14 @@ def to_purl(self):
             protocol = match.group("protocol")
             suffix = match.group("suffix")
             has_authority = match.group("has_authority")
-            if protocol == "file":
-                qualifier = urllib.parse.quote(self.version, safe="")
-                return f"generic/{purl_name}?{qualifier}"
-            elif not has_authority:
+            if not has_authority:
+                if protocol == "file":
+                    path = self.version.split(":", 1)[1]
+                    safe_path = urllib.parse.quote(path, safe="")
+                    # this path is relative to the parent package's package.json path
+                    # let's mark this purl for later substitution
+                    marker = content_manifest.PARENT_PURL_PLACEHOLDER
+                    return f"pkg:generic/{purl_name}?download_url={marker}&file_name={safe_path}"
```

Comment:
Keep in mind that `safe_path` is from the perspective of the `package.json` file of the main package. The main package could be in a subdirectory and thus the relative path of `../client` does not make sense in the PURL.

Reply:
Hmm that's a good point. We should normalize the path then.

Reply:
For this, we would need to know the parent's package.json path, which we do not at this point. How about including the path for the directory containing the package.json in the parent's PURL? Then we could keep the file_name as a relative path (this would likely require changing the version field to include such path).

Reply:
I see. I'm afraid I don't know the full implication of making such a change. Could we normalize `file_name` before it's saved as a dependency instead? That seems like the right thing to do, but it may be more intrusive.